src/expr/src/table_function/mod.rs

```diff
@@ -139,10 +109,76 @@ impl ProjectSetSelectItem {
         }
     }
 
-    pub async fn eval(&self, input: &DataChunk) -> Result<Either<Vec<ArrayRef>, ArrayRef>> {
+    pub async fn eval<'a>(
+        &'a self,
+        input: &'a DataChunk,
+    ) -> Result<Either<TableFunctionOutputIter<'a>, ArrayRef>> {
         match self {
-            ProjectSetSelectItem::TableFunction(tf) => tf.eval(input).await.map(Either::Left),
-            ProjectSetSelectItem::Expr(expr) => expr.eval(input).await.map(Either::Right),
+            Self::TableFunction(tf) => Ok(Either::Left(
+                TableFunctionOutputIter::new(tf.return_type(), tf.eval(input).await).await?,
+            )),
+            Self::Expr(expr) => expr.eval(input).await.map(Either::Right),
         }
     }
 }
+
+/// A wrapper over the output of table function that allows iteration by rows.
+pub struct TableFunctionOutputIter<'a> {
+    data_type: DataType,
+    stream: BoxStream<'a, Result<DataChunk>>,
+    chunk: Option<(ArrayRef, ArrayRef)>,
+    index: usize,
+}
+
+impl<'a> TableFunctionOutputIter<'a> {
+    pub async fn new(
+        data_type: DataType,
+        stream: BoxStream<'a, Result<DataChunk>>,
+    ) -> Result<TableFunctionOutputIter<'a>> {
+        let mut iter = Self {
+            data_type,
+            stream,
+            chunk: None,
+            index: 0,
+        };
+        iter.pop_from_stream().await?;
+        Ok(iter)
+    }
+
+    /// Gets the current row.
+    pub fn peek(&'a self) -> Option<(usize, DatumRef<'a>)> {
+        let (indexes, values) = self.chunk.as_ref()?;
+        let index = indexes.as_int64().value_at(self.index).unwrap() as usize;
+        let value = values.value_at(self.index);
+        Some((index, value))
+    }
+
+    /// Moves to the next row.
+    pub async fn next(&mut self) -> Result<()> {
```

Comment:
This function seems not cancellation-safe. Maybe we should leave a note here and I'm not sure whether it matters now. 🤔

Reply:
I think it's cancellation-safe because the only await point inside this function is `stream.next()`. 🤔

Reply:
If we cancel the call when awaiting the stream, `self.index` should already be updated. 🤔 Then the next call will skip this index.

Maybe we should put `self.index += 1` after `await`?

Reply:
Good catch! Fixed in 90ce5b7.